lucida/commandcenter/controllers/ThriftClient.py

```diff
@@ -10,80 +12,132 @@
-			host, port = service.get_host_port()
-			tag_list = [host, str(port), str(len(node.to_indices))]
-			for to_index in node.to_indices:
-				tag_list.append(str(to_index))
-			query_input_list.append(self.create_query_input(
-				service.input_type, data, tag_list))
-		query_spec = self.create_query_spec('query', query_input_list)
-		# Go through all starting indices and send requests.
-		result = []
-		for start_index in service_graph.starting_indices:
-			service = self.SERVICES[service_graph.get_node(
-				start_index).service_name]
-			client, transport = self.get_client_transport(service)
-			log('Sending infer request to ' + service.name)
-			result.append(client.infer(str(LUCID), query_spec))
-			transport.close()
-		return ' '.join(result)
+    def get_client_transport(self, service):
+        host, port = service.get_host_port()
+        transport = TTransport.TFramedTransport(TSocket.TSocket(host, port))
+        protocol = TBinaryProtocol.TBinaryProtocol(transport)
+        transport.open()
+        return LucidaService.Client(protocol), transport
 
+    def send_query(self, LUCID, service_graph, start_index, query_input_list):
+        query_spec = self.create_query_spec('query', query_input_list)
+        service = self.SERVICES[service_graph.get_node(start_index).service_name]
+        client, transport = self.get_client_transport(service)
+        log('Sending infer request to ' + service.name)
+        result = client.infer(str(LUCID), query_spec)
+        transport.close()
+        return result
 
-thrift_client = ThriftClient(Config.SERVICES)	
+# TODO: If SESSION is within its own file move this function there
+    def clear_context(self, LUCID, service_graph):
+        if LUCID in Config.SESSION:
+            # Initialize workflow back to the beginning
+            service_graph.start_index = 0
+            del Config.SESSION[LUCID]
+
+    def learn_image(self, LUCID, image_type, image_data, image_id):
+        for service in Config.Service.LEARNERS['image']: # add concurrency?
+            knowledge_input = self.create_query_input(
+                image_type, [image_data], [image_id])
+            client, transport = self.get_client_transport(service)
+            log('Sending learn_image request to IMM')
+            client.learn(str(LUCID),
+                self.create_query_spec('knowledge', [knowledge_input]))
+            transport.close()
+
+    def learn_text(self, LUCID, text_type, text_data, text_id):
+        for service in Config.Service.LEARNERS['text']: # add concurrency?
+            knowledge_input = self.create_query_input(
+                text_type, [text_data], [text_id])
+            client, transport = self.get_client_transport(service)
+            log('Sending learn_text request to QA')
+            client.learn(str(LUCID),
+                self.create_query_spec('knowledge', [knowledge_input]))
+            transport.close()
+
+# TODO: I need help trying to make this method look less scary
```

Comment:
make this comment more descriptive what you would want to do (split into multiple functions, make a class) suggesting next that comes here has an idea.